train.py

```diff
@@ -0,0 +1,200 @@
+import os
+
+import hydra
+import torch
+import wandb
+from tqdm import tqdm
+import torch.optim as optim
+import torch.nn.functional as F
+from omegaconf import OmegaConf
+from huggingface_hub import upload_file
+from torch.utils.data import Subset, DataLoader
+
+from models.reverse_diffusion import Unet
+from ecg_segmentation_dataset import ECGDataset
+from models.forward_diffusion import ForwardDiffusion
+
+
+def makedir_if_not_exists(dir: str):
+    if not os.path.exists(dir):
+        os.makedirs(dir)
+
+
+def preprocess_dataset(dataset_name: str, batch_size: int, num_workers: int, *, overfit_single_batch: bool = False):
+    train_ds = ECGDataset(dataset_name, split="train")
+    val_ds = ECGDataset(dataset_name, split="validation")
+    test_ds = ECGDataset(dataset_name, split="test")
+
+    if overfit_single_batch:
+        train_ds = Subset(train_ds, indices=range(batch_size))
+        val_ds = Subset(val_ds, indices=range(batch_size))
+        test_ds = Subset(test_ds, indices=range(batch_size))
+
+    # dataloaders
+    train_dataloader = DataLoader(train_ds, batch_size=batch_size, num_workers=num_workers)
+    val_dataloader = DataLoader(val_ds, batch_size=batch_size, num_workers=num_workers)
+    test_dataloader = DataLoader(test_ds, batch_size=batch_size, num_workers=num_workers)
+
+    return train_dataloader, val_dataloader, test_dataloader
+
+
+def step(
+    model: Unet,
+    forward_diffusion: ForwardDiffusion,
+    batch: dict[str, torch.Tensor, torch.Tensor],
+    device: torch.device,
+    split: str = "train",
+) -> dict:
+    x = batch["signal"].to(device)
+
+    batch_size = x.shape[0]
+
+    # sample t
+    t = torch.randint(0, forward_diffusion.timesteps, size=(batch_size,), dtype=torch.long, device=device)
+
+    # noise batch
+    x_noisy, added_noise = forward_diffusion(x, t)
+
+    # get predicted noise
+    predicted_noise = model(x_noisy, t)
+
+    # get loss value for batch
+    loss = F.mse_loss(predicted_noise, added_noise)
+
+    metrics = {f"{split}/loss": loss}
+
+    return metrics
```

Comment:
```suggestion
    return loss
```

I think it would be better to separate the metrics management from the training steps. 

Also, I think storing tensors with gradients like that creates a risk of "dangling graphs" and memory leaks - which I think I'm running into during the validation step here

Reply:
I think it would be better to return both loss (with grad) and metrics (without grads). When doing it this way training loop is more modular and will work when doing something else, only thing you need to change then is step function.

Reply:
If this step has to return more than just the loss we can have a `dataclass` full of metrics as an output, but I wouldn't spend too much time planning modularity we don't know we need right now :)

For me, the main suggestion here is to remove `split` from this function completely, as it's mixing the model logic with performance tracking logic, which I think should be separate.